src/vos/tests/vts_pm.c

```diff
@@ -2629,6 +2629,116 @@ many_tx(void **state)
 	start_epoch = epoch + 1;
 }
 
+static void
+local_transaction(void **state)
+{
+	struct io_test_args *arg = *state;
+	int                  rc  = 0;
+	int                  passed_rc;
+	daos_key_t           dkey[2];
+	daos_key_t           akey;
+	daos_iod_t           iod;
+	d_sg_list_t          sgl;
+	d_sg_list_t          fetch_sgl;
+	char                 buf[32];
+	char                 buf2[32];
+	daos_epoch_t         epoch = start_epoch;
+	struct dtx_handle   *dth;
+	const char          *first = "Hello";
+	char                 dkey_buf[2][UPDATE_DKEY_SIZE];
+	char                 akey_buf[UPDATE_AKEY_SIZE];
+	daos_unit_oid_t      oid;
+	int                  i;
+
+	test_args_reset(arg, VPOOL_SIZE);
+
+	memset(&iod, 0, sizeof(iod));
+
+	rc = d_sgl_init(&sgl, 1);
+	assert_rc_equal(rc, 0);
+	rc = d_sgl_init(&fetch_sgl, 1);
+	assert_rc_equal(rc, 0);
+
+	/* Set up dkey and akey */
+	oid = gen_oid(arg->otype);
+	for (i = 0; i < 2; i++) {
+		vts_key_gen(&dkey_buf[i][0], arg->dkey_size, true, arg);
+		set_iov(&dkey[i], &dkey_buf[i][0],
+			is_daos_obj_type_set(arg->otype, DAOS_OT_DKEY_UINT64));
+	}
+	vts_key_gen(&akey_buf[0], arg->akey_size, true, arg);
+	set_iov(&akey, &akey_buf[0], is_daos_obj_type_set(arg->otype, DAOS_OT_AKEY_UINT64));
+	iod.iod_type  = DAOS_IOD_SINGLE;
+	iod.iod_size  = strlen(first);
+	iod.iod_name  = akey;
+	iod.iod_recxs = NULL;
+	iod.iod_nr    = 1;
+
+	d_iov_set(&sgl.sg_iovs[0], (void *)first, iod.iod_size);
+
+	for (i = 0; i < 2; i++) {
+		rc = vos_local_tx_begin(arg->ctx.tc_po_hdl, &dth);
+		assert_rc_equal(rc, 0);
+
+		rc = vos_obj_update_ex(arg->ctx.tc_co_hdl, oid, epoch++, 0, 0, &dkey[0], 1, &iod,
+				       NULL, &sgl, dth);
+		assert_rc_equal(rc, 0);
+
+		rc = vos_obj_update_ex(arg->ctx.tc_co_hdl, oid, epoch++, 0, 0, &dkey[1], 1, &iod,
+				       NULL, &sgl, dth);
+		assert_rc_equal(rc, 0);
+
+		rc = vos_obj_punch(arg->ctx.tc_co_hdl, oid, epoch++, 0, 0, &dkey[1], 0, NULL, dth);
+		assert_rc_equal(rc, 0);
+
+		if (i == 0) {
+			/** Abort first time */
+			memset(buf2, 'x', sizeof(buf2));
+			passed_rc = -DER_EXIST;
+		} else {
+			/** Commit second time */
+			passed_rc = 0;
+			memcpy(buf2, first, strlen(first));
+		}
+
+		rc = vos_local_tx_end(dth, passed_rc);
+		assert_rc_equal(rc, passed_rc);
+
```

Comment:
I can't see obvious issue here, what if you comment out the punch, and see if abort(update) can work. Or put punch in a separate transaction and see if we can actually abort(punch).

Reply:
I had a prior version that just did two updates, no punch and it didn't abort at least the first update and subsequent fetch showed the data I wrote.